Dockerfile

```
@@ -40,6 +40,10 @@ COPY package.json ./
 COPY backend/package.json ./backend/
 COPY frontend/package.json ./frontend/
 
+# Install toolchain first for optimal cache: any source change won't re-download gcc
+RUN apt-get update && apt-get install -y --no-install-recommends gcc libc6-dev \
```

Hadolint (2.15.1)

[warning] Pin versions in apt get install. Instead of `apt-get install <package>` use `apt-get install <package>=<version>`

(DL3008)